lunatrace/bsl/backend/src/snapshot/package-tree.ts

```diff
@@ -0,0 +1,165 @@
+ * Licensed under the Business Source License v1.1
+ * (the "License"); you may not use this file except in compliance with the
+ * License. You may obtain a copy of the License at
+ *
+ * https://github.com/lunasec-io/lunasec/blob/master/licenses/BSL-LunaTrace.txt
+ *
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+import { randomUUID } from 'crypto';
+import path from 'path';
+
+import { buildDepTreeFromFiles, PkgTree } from 'nodejs-lockfile-parser';
+import { DepTreeDep } from 'nodejs-lockfile-parser/dist/parsers';
+
+import { hasura } from '../hasura-api';
+import {
+  Build_Dependency_Relationship_Constraint,
+  Build_Dependency_Relationship_Insert_Input,
+  Build_Dependency_Relationship_Update_Column,
+  Package_Constraint,
+  Package_Release_Constraint,
+  Package_Release_Update_Column,
+  Package_Update_Column,
+} from '../hasura-api/generated';
+import { walk } from '../utils/fs';
+import { log } from '../utils/log';
+import { notEmpty } from '../utils/predicates';
+
+export async function collectPackageTreesFromDirectory(repoDir: string) {
+  const lockFileRegex = new RegExp('package-lock\\.json|yarn\\.lock');
+  const lockFilePaths = walk(repoDir, (path, name) => lockFileRegex.test(name));
+
+  return await Promise.all(
+    Array.from(lockFilePaths, async (lockFilePath) => {
+      const { dir, base } = path.parse(lockFilePath);
+      return await buildDepTreeFromFiles(dir, 'package.json', base, true);
+    })
+  );
+}
+
+interface FlattenedDependency {
+  id: string;
+  dependency: DepTreeDep;
+  parentId?: string;
+}
+
+function depDependencies(dep: DepTreeDep, parentId?: string): FlattenedDependency[] {
+  // if these are root dependencies then do not generate a parent identifier, there is none.
+  // this will help identify which are root level dependencies in the database.
+  const id = randomUUID();
+  return Object.values(dep.dependencies || []).map((d) => ({ id, dependency: d, parentId }));
+}
+
+function flattenPkgTree(pkgTree: PkgTree) {
+  const stack = [...depDependencies(pkgTree)];
+  const res = [];
+  while (stack.length) {
+    // pop value from stack
+    const next = stack.pop();
+    if (!next) {
+      break;
+    }
+
+    stack.push(...depDependencies(next.dependency, next.id));
+    res.push(next);
+  }
+  // reverse to restore input order
+  return res;
+}
+
+function mapPackageTreeToBuildDependencyRelationships(
+  buildId: string,
+  pkgTree: PkgTree
+): Build_Dependency_Relationship_Insert_Input[] {
+  const flattenedDependencies = flattenPkgTree(pkgTree);
+  return flattenedDependencies
+    .map((entry): Build_Dependency_Relationship_Insert_Input | null => {
+      const { id, dependency, parentId } = entry;
+      if (!dependency.version || !dependency.name) {
+        return null;
+      }
+
+      return {
+        /*
+         TODO (cthompson) we are generating the ids for the database here. This avoids a roundtrip from the server
```

Comment:
You can delete this comment
> [As per Wikipedia](https://en.wikipedia.org/wiki/Universally_unique_identifier), the number of UUIDs generated to have atleast 1 collision is 2.71 quintillion.
> 
> This is equivalent to generating around 1 billion UUIDs per second for about 85 years. Oh Boy!! that’s a really long time.
> 
> 
Basically, we never worry about this actually happening. 

Another option, to avoid generating these UUIDs on the client out of laziness and concern about bugs/complexity, would be to just build the entire tree in the correct shape for hasura, in memory, and send it up in that shape.  Hasura figures out the relationships and builds the foreign keys like a good ORM should :) I understand that's a bit of work to get it into that structure since the manifest files are flat and things don't really get parsed in order.  Just another option I wanted to mention.